File: src/shared/Contracts/Skill/Requests/CreateSkillRequest.cs
using System.ComponentModel.DataAnnotations;
using Contracts.Common;

namespace Contracts.Skill.Requests;

/// <summary>
/// API request for creating a new skill
/// </summary>
public record CreateSkillRequest(
    [Required(ErrorMessage = "Skill name is required")]
    [StringLength(100, MinimumLength = 2, ErrorMessage = "Skill name must be between 2 and 100 characters")]
    string Name,

    [Required(ErrorMessage = "Description is required")]
    [StringLength(2000, MinimumLength = 10, ErrorMessage = "Description must be between 10 and 2000 characters")]
    string Description,

    [Required(ErrorMessage = "Category ID is required")]
    string CategoryId,

    List<string> Tags,

    [Required(ErrorMessage = "Must specify if skill is offered")]
    bool IsOffered,

    // Legacy fields (optional)
    [Range(1, 40, ErrorMessage = "Available hours must be between 1 and 40 per week")]
    int? AvailableHours = null,

    [Range(15, 480, ErrorMessage = "Session duration must be between 15 and 480 minutes")]
    int? PreferredSessionDuration = 60,

    // ==========================================================================
    // EXCHANGE OPTIONS
    // ==========================================================================

    /// <summary>
    /// Type of exchange: 'skill_exchange' (default) or 'payment'
    /// </summary>
    string? ExchangeType = "skill_exchange",

    /// <summary>
    /// For skill_exchange: Category of the desired skill in return
    /// </summary>
    string? DesiredSkillCategoryId = null,

    /// <summary>
    /// For skill_exchange: Description of what skill the user wants in return
    /// </summary>
    [StringLength(500, ErrorMessage = "Desired skill description cannot exceed 500 characters")]
    string? DesiredSkillDescription = null,

    /// <summary>
    /// For payment: Hourly rate (only when IsOffered=true)
    /// </summary>
    [Range(5, 500, ErrorMessage = "Hourly rate must be between 5 and 500")]
    decimal? HourlyRate = null,

    /// <summary>
    /// For payment: Currency code (EUR, USD, CHF, GBP)
    /// </summary>
    [StringLength(3, ErrorMessage = "Currency must be 3 characters")]
    string? Currency = "EUR",

    // ==========================================================================
    // SCHEDULING
    // ==========================================================================

    /// <summary>
    /// Preferred days of the week (monday, tuesday, etc.)
    /// </summary>
    List<string>? PreferredDays = null,

    /// <summary>
    /// Preferred times of day (morning, afternoon, evening)
    /// </summary>
    List<string>? PreferredTimes = null,

    /// <summary>
    /// Duration of each session in minutes (15, 30, 45, 60, 90, 120)
    /// </summary>
    [Range(15, 120, ErrorMessage = "Session duration must be between 15 and 120 minutes")]
    int SessionDurationMinutes = 60,

    /// <summary>
    /// Total number of sessions
    /// </summary>
    [Range(1, 50, ErrorMessage = "Total sessions must be between 1 and 50")]
    int TotalSessions = 1,

    // ==========================================================================
    // LOCATION
    // ==========================================================================

    /// <summary>
    /// Location type: 'remote' (default), 'in_person', or 'both'
    /// </summary>
    string? LocationType = "remote",

    /// <summary>
    /// Street address (for in_person or both)
    /// </summary>
    [StringLength(200, ErrorMessage = "Address cannot exceed 200 characters")]
    string? LocationAddress = null,

    /// <summary>
    /// City name
    /// </summary>
    [StringLength(100, ErrorMessage = "City name cannot exceed 100 characters")]
    string? LocationCity = null,

    /// <summary>
    /// Postal code
    /// </summary>
    [StringLength(20, ErrorMessage = "Postal code cannot exceed 20 characters")]
    string? LocationPostalCode = null,

    /// <summary>
    /// Country code (ISO 3166-1 alpha-2, e.g., "DE")
    /// </summary>
    [StringLength(2, ErrorMessage = "Country must be 2-letter ISO code")]
    string? LocationCountry = null,

    /// <summary>
    /// Maximum distance in km for in-person meetings
    /// </summary>
    [Range(1, 500, ErrorMessage = "Max distance must be between 1 and 500 km")]
    int MaxDistanceKm = 50)
    : IVersionedContract
{
    /// <summary>
    /// API Version this request supports
    /// </summary>
    public string ApiVersion => "v1";
}

File: src/client/src/features/skills/types/CreateSkillRequest.ts
/**
 * Create Skill Request
 *
 * Extended with Exchange, Scheduling, and Location options.
 * NOTE: ExchangeType "free" is NOT available for skills - only skill_exchange and payment.
 * Public Workshops (free multi-call) are a separate feature (see docs/architecture/public-workshop-epic.md)
 */
export interface CreateSkillRequest {
  // Basic fields (required)
  name: string;
  description: string;
  isOffered: boolean; // true = offering/teaching, false = seeking/learning
  categoryId: string;
  tags?: string[];

  // ==========================================================================
  // IMAGE OPTIONS
  // ==========================================================================

  /**
   * Image option for the skill
   * - 'none': No image, show category letter (default)
   * - 'profile': Use the user's profile photo
   * - 'upload': Upload a new custom image
   */
  imageOption?: 'none' | 'profile' | 'upload';

  /**
   * For upload: Base64-encoded image data
   */
  imageData?: string;

  /**
   * For upload: Original filename
   */
  imageFileName?: string;

  /**
   * For upload: Content type (e.g., 'image/jpeg', 'image/png')
   */
  imageContentType?: string;

  // Legacy fields (optional, for backward compatibility)
  availableHours?: number;
  preferredSessionDuration?: number;

  // ==========================================================================
  // EXCHANGE OPTIONS (optional, default: skill_exchange)
  // ==========================================================================

  /**
   * Type of exchange for this skill
   * - 'skill_exchange': Trade skills with another user
   * - 'payment': Accept payment for teaching (only when isOffered=true)
   *
   * NOTE: 'free' is NOT an option for skills. Public Workshops are a separate feature.
   */
  exchangeType?: 'skill_exchange' | 'payment';

  /**
   * For skill_exchange: Category of the skill you want in return
   */
  desiredSkillCategoryId?: string;

  /**
   * For skill_exchange: Description of what skill you're looking for
   */
  desiredSkillDescription?: string;

  /**
   * For skill_exchange when seeking (isOffered=false): ID of own skill to offer in exchange
   * This allows users to select one of their own offered skills as a counter-offer
   */
  offeredSkillId?: string;

  /**
   * For payment: Hourly rate (only valid when isOffered=true and exchangeType='payment')
   * Min: 5, Max: 500
   */
  hourlyRate?: number;

  /**
   * For payment: Currency code (EUR, USD, CHF, GBP)
   */
  currency?: string;

  // ==========================================================================
  // SCHEDULING (required for matching)
  // ==========================================================================

  /**
   * Preferred days of the week
   * Values: 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday'
   */
  preferredDays?: string[];

  /**
   * Preferred times of day
   * Values: 'morning', 'afternoon', 'evening'
   */
  preferredTimes?: string[];

  /**
   * Duration of each session in minutes
   * Valid values: 15, 30, 45, 60, 90, 120
   * Default: 60
   */
  sessionDurationMinutes?: number;

  /**
   * Total number of sessions to complete the skill exchange
   * Min: 1, Max: 50
   * Default: 1
   */
  totalSessions?: number;

  // ==========================================================================
  // LOCATION (optional, default: remote)
  // ==========================================================================

  /**
   * Location type for skill exchange
   * - 'remote': Online only (default)
   * - 'in_person': In-person only (requires city + country)
   * - 'both': Either remote or in-person
   */
  locationType?: 'remote' | 'in_person' | 'both';

  /**
   * Street address (for in_person or both)
   */
  locationAddress?: string;

  /**
   * City name (required for in_person or both)
   */
  locationCity?: string;

  /**
   * Postal code (for in_person or both)
   */
  locationPostalCode?: string;

  /**
   * Country code (ISO 3166-1 alpha-2, e.g., 'DE', 'AT', 'CH')
   * Required for in_person or both
   */
  locationCountry?: string;

  /**
   * Maximum distance in km for in-person meetings
   * Default: 50, Min: 1, Max: 500
   */
  maxDistanceKm?: number;
}

// =============================================================================
// TYPE GUARDS AND HELPERS
// =============================================================================

/**
 * Valid exchange types for skills (no 'free' - that's for Public Workshops)
 */
export const SKILL_EXCHANGE_TYPES = ['skill_exchange', 'payment'] as const;
export type SkillExchangeType = (typeof SKILL_EXCHANGE_TYPES)[number];

/**
 * Valid location types
 */
export const SKILL_LOCATION_TYPES = ['remote', 'in_person', 'both'] as const;
export type SkillLocationType = (typeof SKILL_LOCATION_TYPES)[number];

/**
 * Valid session durations in minutes
 */
export const SKILL_SESSION_DURATIONS = [15, 30, 45, 60, 90, 120] as const;
export type SkillSessionDuration = (typeof SKILL_SESSION_DURATIONS)[number];

/**
 * Valid weekdays
 */
export const SKILL_WEEKDAYS = [
  'monday',
  'tuesday',
  'wednesday',
  'thursday',
  'friday',
  'saturday',
  'sunday',
] as const;
export type SkillWeekday = (typeof SKILL_WEEKDAYS)[number];

/**
 * Valid time slots
 */
export const SKILL_TIME_SLOTS = ['morning', 'afternoon', 'evening'] as const;
export type SkillTimeSlot = (typeof SKILL_TIME_SLOTS)[number];

/**
 * Check if exchange type requires payment fields
 */
export const isPaymentExchange = (exchangeType?: string): exchangeType is 'payment' =>
  exchangeType === 'payment';

/**
 * Check if location type requires address fields
 */
export const requiresLocationDetails = (
  locationType?: string
): locationType is 'in_person' | 'both' => locationType === 'in_person' || locationType === 'both';
